cmd/satellite/gracefulexit.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package main
+
+import (
+	"context"
+	"encoding/csv"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+
+	"storj.io/storj/satellite/satellitedb"
+)
+
+// generateGracefulExitCSV creates a report with graceful exit data for exiting or exited nodes in a given period
+func generateGracefulExitCSV(ctx context.Context, start time.Time, end time.Time, output io.Writer) error {
+	db, err := satellitedb.New(zap.L().Named("db"), gracefulExitCfg.Database)
+	if err != nil {
+		return errs.New("error connecting to master database on satellite: %+v", err)
+	}
+	defer func() {
+		err = errs.Combine(err, db.Close())
+	}()
+
+	// TODO: add overlay function for getting exiting/exited nodes in timespan
+	// TODO: add a config param similar to storagenode/retain config.Status for choosing between all/exiting/exited
+	nodeIDs, err := db.OverlayCache().GetGracefulExitNodes(ctx, false, start, end)
+	if err != nil {
+		return err
+	}
+
+	w := csv.NewWriter(output)
+	headers := []string{
+		"nodeID",
+		"walletAddress",
+		"nodeCreationDate",
+		"initiatedGracefulExit",
+		"completedGracefulExit",
+		"transferredGB",
+	}
+	if err := w.Write(headers); err != nil {
+		return err
+	}
+
+	for _, id := range nodeIDs {
```

Comment:
Why do we get a list of nodes, then query for each node, instead of doing a join to get all the information in one swoop (via a JOIN or whatever)?  This seems inefficient.

Reply:
A couple reasons:
1. We need data from both the "overlay cache db" and the "graceful exit db", and I don't know of a precedent for joining tables across different dbs. I don't want to make the overlay cache even more confusing than it already is.
2. The query would have to be pretty large to include not only the normal node dossier fields, but also the exit status and exit progress structs.

Maybe these aren't good reasons though.

Reply:
"can't join between databases" is a very good reason.